target/linux/generic/hack-6.12/999-ledtrig.patch

```diff
@@ -0,0 +1,423 @@
++            cancel_delayed_work_sync(&d->work);
++            schedule_delayed_work(&d->work, 0);
++            break;
++        }
++    }
++    spin_unlock_bh(&d->lock);
++    return NOTIFY_DONE;
++}
++
++static int lan_activate(struct led_classdev *led_cdev)
++{
++    struct led_netdev_lan *d;
++
++    d = kzalloc(sizeof(*d), GFP_KERNEL);
++    if (!d)
++        return -ENOMEM;
++
++    spin_lock_init(&d->lock);
++    INIT_DELAYED_WORK(&d->work, lan_work_fn);
++    d->notifier.notifier_call = lan_notify;
++    d->notifier.priority = 0;
++    d->led_cdev = led_cdev;
++    atomic_set(&d->interval, msecs_to_jiffies(DEFAULT_INTERVAL));
++    d->last_rx_sum = 0;
++    d->last_tx_sum = 0;
++
++    /* Initial detection will happen during first work run */
++    d->iface_count = 0;
++
++    register_netdevice_notifier(&d->notifier);
++    schedule_delayed_work(&d->work, 0);
++    led_cdev->trigger_data = d;
++    return 0;
++}
++
++static void lan_deactivate(struct led_classdev *led_cdev)
++{
++    struct led_netdev_lan *d = led_cdev->trigger_data;
++    int i;
++
++    unregister_netdevice_notifier(&d->notifier);
++    cancel_delayed_work_sync(&d->work);
++    for (i = 0; i < d->iface_count; i++)
++        dev_put(d->netdevs[i]);
++
++    led_set_brightness(led_cdev, LED_OFF);
++    kfree(d);
++}
++
++static struct led_trigger netdev_lan_trigger = {
++    .name = "netdev_lan",
++    .activate = lan_activate,
++    .deactivate = lan_deactivate,
++};
++
++module_led_trigger(netdev_lan_trigger);
++
++MODULE_AUTHOR("Mieczyslaw Nalewaj namiltd@yahoo.com");
++MODULE_DESCRIPTION("Netdev LED trigger for up to 16 lan* interfaces with RX/TX sumation and OR logic");
++MODULE_LICENSE("GPL v2");
+--- /dev/null
++++ b/drivers/leds/trigger/ledtrig-netdev-phy.c
+@@ -0,0 +1,203 @@
++// SPDX-License-Identifier: GPL-2.0
++// LED trigger: emulate mac80211 tpt logic with blink at all traffic levels, but solid ON only when no traffic.
++
++#include <linux/atomic.h>
++#include <linux/jiffies.h>
++#include <linux/leds.h>
++#include <linux/module.h>
++#include <linux/netdevice.h>
++#include <linux/rtnetlink.h>
++#include <linux/spinlock.h>
++#include <linux/workqueue.h>
++#include "../leds.h"
++
++#define MAX_IFACES 16
++#define FIXED_INTERVAL_MS 100
++
++/* Blink timing table (kbps thresholds, on/off ms) */
++static const struct {
++    u32 throughput;
++    long unsigned int on_ms;
++    long unsigned int off_ms;
++} tpt_table[] = {
++    {  64,   200, 800 },
++    { 512,   200, 300 },
++    {2048,   200, 150 },
++    {10000,  200,  50 },
++    {54000,  100,  50 }, /* now blink instead of solid ON */
++};
++
++struct led_netdev_phy {
++    spinlock_t lock;
++    struct delayed_work work;
++    struct notifier_block notifier;
++    struct led_classdev *led_cdev;
++    struct net_device *netdevs[MAX_IFACES];
++    int iface_count;
++
++    u64 last_rx_bytes_sum;
++    u64 last_tx_bytes_sum;
++};
++
++/* Detect all interfaces with names starting with "phy" */
```

Comment:
This does not sound future proof due to wiphy renaming.
Eventually several targets will use renaming so we don't need migration scripts anymore (for existing uci config) if pcie paths change due to kernel updates.
The only target currently doing this in release is mediatek-mt7622.
phy0 is renamed to wl0
phy1 is renamed to wl1
(depending on the situation it could also be phy1 -> wl0, phy0 -> wl1; renaming allows consistent numbering like 2.4GHz being the first phy etc.)
Right now renaming only happens once an interface goes up not before, which might change in the future.

Some more info and the original commit are mentioned here:
https://github.com/openwrt/openwrt/pull/17821

Reply:
This is just a draft, there is no problem in adding "wl*" to the code as well.